File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/DropoutPerform.scala
package com.intel.webscaleml.nn.nn.performTest

import java.io.{BufferedWriter, File, FileWriter}

import com.intel.webscaleml.nn.nn.Dropout
import com.intel.webscaleml.nn.tensor.{torch}
import org.scalatest.{BeforeAndAfter, FlatSpec}
import com.intel.webscaleml.nn.tensor.RandomGenerator._

/**
  * Created by yao on 5/19/16.
  */
class DropoutPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val InputNum = 1000
  val FeatureDim = 512
  val Seed = 100
  val allTestCases = List(
    TestCase(0.3, 42, 25),
    TestCase(0.4, 45, 25),
    TestCase(0.5, 43, 25)
  )

  RNG.setSeed(Seed);
  val input = torch.Tensor[Float](InputNum, FeatureDim).rand()

  for (testCase <- allTestCases){
    val model = new Dropout[Float](testCase.p)
    "A Dropout" should s" forward with a good performance with parameters: ${testCase.p}" in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")

      val timeMillis = TestUtils.testModuleForwardPerform(input, model, ForwardIterations)
      //assert(timeMillis < testCase.ForwardTimeoutMillis)
    }
  }

  for (testCase <- allTestCases) {
    val model = new Dropout[Float](testCase.p)
    val output = model.forward(input)
    val grads = output.clone().rand()
    it should s" backward with a good performance with parameters: ${testCase.p}" in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")

      val timeMillis = TestUtils.testModuleBackwardPerform(input, grads, model, BackwardIterations)
      //assert(timeMillis < testCase.BackwardTimeoutMillis)
    }
  }

  case class TestCase(p: Double, ForwardTimeoutMillis: Int, BackwardTimeoutMillis: Int)
}

File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/DropoutPerform.lua
--
-- Created by yansh on 16-6-22.
--

require 'nn'

forwardIterations = 10
backwardIterations = 10
inputNum = 1000
featureDim = 512
seed = 100
allTestCases = {
    {p = 0.3, forwardTimeoutMillis = 32, backwardTimeoutMillis = 19},
    {p = 0.4, forwardTimeoutMillis = 35, backwardTimeoutMillis = 20},
    {p = 0.5, forwardTimeoutMillis = 32, backwardTimeoutMillis = 20}
}

torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)

input = torch.Tensor(inputNum, featureDim)

for _, testCase in pairs(allTestCases) do
    local model = nn.Dropout(testCase.p)
    local timer = torch.Timer()
    for j = 1, forwardIterations do
        model:forward(input)
    end
    local timeSpend = timer:time().real / forwardIterations
    file:write("Dropout(" .. testCase.p ..  ")forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

for _, testCase in pairs(allTestCases) do
    local model = nn.Dropout(testCase.p)
    local output = model:forward(input)
    local grads = output:clone()
    local timer = torch.Timer()
    for j = 1, backwardIterations do
        model:backward(input, grads)
    end
    local timeSpend = timer:time().real / backwardIterations
    file:write("Dropout(" .. testCase.p .. ")backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

file:close()
